### ISSUE TITLE
Query method for `hint` does not support named indexes

### ISSUE THREAD

**Issue by JasonCust** [Automattic/mongoose#4934]
Tested on Node 6.9.4, MongoDB: 3.4, MongooseJS 4.7.8

The [MongoDB documentation for `hint`](https://docs.mongodb.com/manual/reference/method/cursor.hint/#cursor.hint) allows the caller to "[s]pecify the index either by the index name or by the index specification document." However when attempting to do this with the [MongooseJS `hint` method](http://mongoosejs.com/docs/api.html#query_Query-hint) it currently only supports the index specification document.

When attempting to use a named index as the argument the following error is returned: `TypeError: Invalid hint. [NAME OF INDEX SPECIFIED]`

This [node driver bug](https://jira.mongodb.org/browse/NODE-89?jql=project%20%3D%20NODE%20AND%20text%20~%20%22hint%20name%22) might be related.

To reproduce:

```javascript
/* jshint node:true, esversion:6 */
/* jscs:disable requireTrailingComma */

const mongoose = require('mongoose');
const explain = require('mongoose-explain');

// Set internal promise object to be the native Promise object
mongoose.Promise = global.Promise;
mongoose.set('debug', true);

const connection = mongoose.createConnection('mongodb://localhost:32768/hint_test');

connection.once('connected', function () {
  const PersonSchema = new mongoose.Schema({
    name: {
      family: { type: String, trim: true },
      given: { type: String, trim: true }
    }
  });

  PersonSchema.index({
    'name.family': 1,
    'name.given': 1
  }, {
    name: 'my_named_index',
    background: false
  });

  PersonSchema.plugin(explain);

  const Person = connection.model('Person', PersonSchema);

  Person.on('index', function (error) {
    Person.insertMany([
      { name: {
        given: 'Luke',
        family: 'Skywalker'
      } },
      { name: {
        given: 'Han',
        family: 'Solo'
      } },
      { name: {
        given: 'Leia',
        family: 'Organa'
      } },
      { name: {
        given: 'Lando',
        family: 'Calrissian'
      } },
      { name: {
        given: 'Tendra',
        family: 'Risant'
      } }
    ]).then(function () {
      Person.collection.getIndexes().then(function (indexes) {
        Person.find({
          'name.family': 'Skywalker'
        }).exec().then(function () {
          Person.find({
            'name.family': 'Solo'
          }).hint({
            'name.family': 1,
            'name.given': 1
          }).exec().then(function () {
            Person.find({
              'name.family': 'Organa'
            }).hint('my_named_index').exec();
          });
        });
      });
    });
  });
});
```
Will log the following:
```shell
Mongoose: people.ensureIndex({ 'name.family': 1, 'name.given': 1 }, { name: 'my_named_index', background: false })
Mongoose: people.insertMany([ { __v: 0, _id: 588a47487ad7b54a310557be, name: { given: 'Luke', family: 'Skywalker' } }, { __v: 0, _id: 588a47487ad7b54a310557bf, name: { given: 'Han', family: 'Solo' } }, { __v: 0, _id: 588a47487ad7b54a310557c0, name: { given: 'Leia', family: 'Organa' } }, { __v: 0, _id: 588a47487ad7b54a310557c1, name: { given: 'Lando', family: 'Calrissian' } }, { __v: 0, _id: 588a47487ad7b54a310557c2, name: { given: 'Tendra', family: 'Risant' } } ])
Mongoose: people.indexInformation()
Mongoose: people.find({ 'name.family': 'Skywalker' }, { fields: undefined })
Mongoose: people.find({ 'name.family': 'Skywalker' }, { explain: true, fields: undefined })
[ { queryPlanner: 
     { plannerVersion: 1,
       namespace: 'hint_test.people',
       indexFilterSet: false,
       parsedQuery: { 'name.family': { '$eq': 'Skywalker' } },
       winningPlan: 
        { stage: 'FETCH',
          inputStage: 
           { stage: 'IXSCAN',
             keyPattern: { 'name.family': 1, 'name.given': 1 },
             indexName: 'my_named_index',
             isMultiKey: false,
             multiKeyPaths: { 'name.family': [], 'name.given': [] },
             isUnique: false,
             isSparse: false,
             isPartial: false,
             indexVersion: 2,
             direction: 'forward',
             indexBounds: 
              { 'name.family': [ '["Skywalker", "Skywalker"]' ],
                'name.given': [ '[MinKey, MaxKey]' ] } } },
       rejectedPlans: [] },
    executionStats: 
     { executionSuccess: true,
       nReturned: 1,
       executionTimeMillis: 0,
       totalKeysExamined: 1,
       totalDocsExamined: 1,
       executionStages: 
        { stage: 'FETCH',
          nReturned: 1,
          executionTimeMillisEstimate: 0,
          works: 2,
          advanced: 1,
          needTime: 0,
          needYield: 0,
          saveState: 0,
          restoreState: 0,
          isEOF: 1,
          invalidates: 0,
          docsExamined: 1,
          alreadyHasObj: 0,
          inputStage: 
           { stage: 'IXSCAN',
             nReturned: 1,
             executionTimeMillisEstimate: 0,
             works: 2,
             advanced: 1,
             needTime: 0,
             needYield: 0,
             saveState: 0,
             restoreState: 0,
             isEOF: 1,
             invalidates: 0,
             keyPattern: { 'name.family': 1, 'name.given': 1 },
             indexName: 'my_named_index',
             isMultiKey: false,
             multiKeyPaths: { 'name.family': [], 'name.given': [] },
             isUnique: false,
             isSparse: false,
             isPartial: false,
             indexVersion: 2,
             direction: 'forward',
             indexBounds: 
              { 'name.family': [ '["Skywalker", "Skywalker"]' ],
                'name.given': [ '[MinKey, MaxKey]' ] },
             keysExamined: 1,
             seeks: 1,
             dupsTested: 0,
             dupsDropped: 0,
             seenInvalidated: 0 } },
       allPlansExecution: [] },
    serverInfo: 
     { host: 'b73ba9db306b',
       port: 27017,
       version: '3.4.1',
       gitVersion: '5e103c4f5583e2566a45d740225dc250baacfbd7' },
    ok: 1 } ]
Mongoose: people.find({ 'name.family': 'Solo' }, { hint: { 'name.family': 1, 'name.given': 1 }, fields: undefined })
Mongoose: people.find({ 'name.family': 'Solo' }, { explain: true, fields: undefined })
[ { queryPlanner: 
     { plannerVersion: 1,
       namespace: 'hint_test.people',
       indexFilterSet: false,
       parsedQuery: { 'name.family': { '$eq': 'Solo' } },
       winningPlan: 
        { stage: 'FETCH',
          inputStage: 
           { stage: 'IXSCAN',
             keyPattern: { 'name.family': 1, 'name.given': 1 },
             indexName: 'my_named_index',
             isMultiKey: false,
             multiKeyPaths: { 'name.family': [], 'name.given': [] },
             isUnique: false,
             isSparse: false,
             isPartial: false,
             indexVersion: 2,
             direction: 'forward',
             indexBounds: 
              { 'name.family': [ '["Solo", "Solo"]' ],
                'name.given': [ '[MinKey, MaxKey]' ] } } },
       rejectedPlans: [] },
    executionStats: 
     { executionSuccess: true,
       nReturned: 1,
       executionTimeMillis: 0,
       totalKeysExamined: 1,
       totalDocsExamined: 1,
       executionStages: 
        { stage: 'FETCH',
          nReturned: 1,
          executionTimeMillisEstimate: 0,
          works: 2,
          advanced: 1,
          needTime: 0,
          needYield: 0,
          saveState: 0,
          restoreState: 0,
          isEOF: 1,
          invalidates: 0,
          docsExamined: 1,
          alreadyHasObj: 0,
          inputStage: 
           { stage: 'IXSCAN',
             nReturned: 1,
             executionTimeMillisEstimate: 0,
             works: 2,
             advanced: 1,
             needTime: 0,
             needYield: 0,
             saveState: 0,
             restoreState: 0,
             isEOF: 1,
             invalidates: 0,
             keyPattern: { 'name.family': 1, 'name.given': 1 },
             indexName: 'my_named_index',
             isMultiKey: false,
             multiKeyPaths: { 'name.family': [], 'name.given': [] },
             isUnique: false,
             isSparse: false,
             isPartial: false,
             indexVersion: 2,
             direction: 'forward',
             indexBounds: 
              { 'name.family': [ '["Solo", "Solo"]' ],
                'name.given': [ '[MinKey, MaxKey]' ] },
             keysExamined: 1,
             seeks: 1,
             dupsTested: 0,
             dupsDropped: 0,
             seenInvalidated: 0 } },
       allPlansExecution: [] },
    serverInfo: 
     { host: 'b73ba9db306b',
       port: 27017,
       version: '3.4.1',
       gitVersion: '5e103c4f5583e2566a45d740225dc250baacfbd7' },
    ok: 1 } ]
(node:18993) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: Invalid hint. my_named_index
```

**Comment by vkarpov15** [comment]
Bug with mquery, the [node driver handles this just fine](https://github.com/mongodb/node-mongodb-native/blob/31bcf9764dbd9dcada50185fe2936e1cf89aef36/test/functional/cursor_tests.js#L2502). Will be fixed in 4.8.1 :+1: :rocket:

**Comment by jrthib** [comment]
Thanks!
